### ISSUE TITLE
[css-shapes] Consider adding shape-outside: line-box

### ISSUE THREAD

**Issue by jpamental** [w3c/csswg-drafts#4003]
I'd like to propose adding a new value to the `shape-outside` property of `line-box`. This would enable floating a title or other block of text and allowing text to wrap around it following the contour of that floated text. It would operate in a similar manner to `shape-outside: <image>` when the image has an alpha-channel transparency. It should also leverage `shape-margin` attributes as well.

While it's possible to achieve this using a `polygon` shape, adding `line-box` would mean it would work with any text rather than having to be custom-coded.

**Examples**
<img width="701" alt="Screen Shot 2019-06-04 at 1 32 19 PM" src="https://user-images.githubusercontent.com/537413/58900763-c8652400-86cd-11e9-9c04-a153de010dc5.png">

Here is a CodePen example using `shape-outside: <polygon>` to mimic the effect:
https://codepen.io/jpamental/pen/LovYov

Here is a real-world example that is a bit more responsive:
https://rwt.io/blog/2018/10/georgia-my-mind

**Shape-outside spec**
https://drafts.csswg.org/css-shapes/#shape-outside-property

**Comment by mirisuzanne** [comment]
This would also be useful as a clip-path, making it easier to create a reliable [ragged background effect](https://css-tricks.com/multi-line-padded-text/).

**Comment by astearns** [comment]
It might be useful to add something to https://drafts.csswg.org/css-inline-3/#inline-box-dimensions for the ragged background effect. It currently defines the block-oriented sizing of the line box's content area. We could add inline-oriented sizing controls there as well (which might then affect the clip-path and shape-outside bounds for the value proposed here)

**Comment by MatsPalmgren** [comment]
Here's an example from a printed magazine.  In this case, it's not just the line box that defines the shape but the outline of the text glyph:
![image](https://user-images.githubusercontent.com/4010828/97111264-2ab52f00-16de-11eb-85f7-db66e8ff43ff.png)

**Comment by faceless2** [comment]
@MatsPalmgren you could do that one with `initial-letter` already - no need for a change to `shape-outside`.

**Comment by MatsPalmgren** [comment]
@faceless2 Oh, I didn't know that. You're right, it looks like [initial-letter-wrap](https://drafts.csswg.org/css-inline/#propdef-initial-letter-wrap) address that use case.

**Comment by jpamental** [comment]
Thanks for pointing to that one @faceless2 - now if only we could get browsers to implement initial-letter. So far Safari is the only one, and only partially at that :(

**Comment by MatsPalmgren** [comment]
Perhaps it could be generalized to all floats by adding something like that to `shape-outside` though? I suspect that if you implement `initial-letter-wrap:all` then it would be trivial to re-use the same code to apply it to floats.  That said, I don't know if it would be useful beyond first-letters...

**Comment by faceless2** [comment]
I agree there's a definite requirement here - `initial-letter-wrap: all` isn't going to help if you've got more than one line.

To me it this looks like yet another place where the requirement is "take a rendered DOM fragment and use the rendering as input to the layout or rendering of another DOM element". I listed a few of these in https://github.com/w3c/csswg-drafts/issues/1981, and as I suggested there, I'd personally like to see all these consolidated under the `url()` function.

This would mean `url(#foo)` means "the bitmap image of the stacking context #foo", when we want to treat it as an image (i.e. if referenced as a background-image or mask-image), or "the outline of the stacking context #foo" (if referenced as a clip-path or shape-outside)

In other words, I don't see any reason why you couldn't do this:

```html
<style>
#floated {
    shape-outside: url(#floated);
    float: left;
    width: 10em;
}
</style>
<div id="floated">Such a very strange odyssey indeed</div>
```

shape-outside doesn't affect the layout of the element it's set on, so there's no dependency loop here.

**Comment by fantasai** [comment]
@faceless2 But why do we need the ID in that case? It's a self-reference, couldn't we make it a keyword so that you can set it on a set of elements at once?

**Comment by faceless2** [comment]
For this individual case? Sure, great idea - `shape-outside: self` or something. But I'm still angling for a more general solution where a URL to a stacking-context-root element can be used to retrieve its outline (for clipping) and its rendered bitmap (for everything else). `shape-outside: self` is just a special case of that.
